### DISCUSSION TITLE
Swagger response name very long

### DISCUSSION THREAD

**Discussion by shurshilov** [fastapi/fastapi#11316] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@router.get(
    "/api/show_proxy_option_value_types/",
    response_model=list[ShowProxyOptionValueTypeItemResponse],
)
async def show_proxy_option_value_types(req: Request):
   ...
```

### Description

Hi!
Name of response model its summary python function "show_proxy_option_value_types" name plus "/api/show_proxy_option_value_types/" its very long. Attach screenshot...
We use RTKQ(typescript) code generation and the hooks on the frontend will turn out to be too long.
Can you tell me why this is so?

### Operating System

Linux

### Operating System Details

docker python-3.11

### FastAPI Version

0.109.2

### Pydantic Version

2.5.3

### Python Version

3.11

### Additional Context

![image](https://github.com/tiangolo/fastapi/assets/11828278/e78bda8d-1923-4683-aba9-c4df8fbbbcb9)

**Comment by YuriiMotov** [comment, marked as answer]
You can change these ids by using custom function for generating unique ids:
```
import re

from fastapi import FastAPI
from fastapi.routing import APIRoute
from pydantic import BaseModel


def my_uniq_id_generate(route: APIRoute) -> str:
    operation_id = f"{route.name}{route.path_format}"
    operation_id = re.sub(r"\W", "_", operation_id)
    assert route.methods
    operation_id = f"{operation_id}_{list(route.methods)[0].lower()}"
    return operation_id

app = FastAPI(
    generate_unique_id_function=my_uniq_id_generate
)


class ShowProxyOptionValueTypeItemResponse(BaseModel):
    s: int


@app.get(
    "/api/show_proxy_option_value_types/",
    response_model=list[ShowProxyOptionValueTypeItemResponse]
)
def show_proxy_option_value_types():
    pass
```

Here `my_uniq_id_generate` is just a copy of default function. You can customize it as you need

**Reply by shurshilov** [reply]
Thanks, I do not know why fastcapi does this and how it generates the operation_id. But to switch from aiohttp to fastapi and replace aiohttp-swagger, I had to delete this field and everything worked as it should.
```
def custom_openapi():
    if app.openapi_schema:
        return app.openapi_schema
    openapi_schema = get_openapi(
        title=TITLE,
        description=DESCRIPTION,
        version="3.0.0",
        routes=app.routes,
    )

    for route in openapi_schema["paths"].keys():
        path = openapi_schema["paths"][route]
        method = (
            path.get("get") or path.get("post") or path.get("put") or path.get("delete")
        )
        if method:
            del method["operationId"]
    app.openapi_schema = openapi_schema
    return app.openapi_schema
app.openapi = custom_openapi
```

**Comment by KunxiSun** [comment]
I met the same problem when I use union type. Openapi generate a very long name for the code below:

```python 

LongNameClass = Annotated[
    LongNameClassA | LongNameClassB, Field(discriminator="type")
]

@app.get(
    "/api/my",
    response_model=LongNameClass,
)
def my_api():
    pass

```

Then, I find a better way to customize the name. Use `pydantic.RootModel` can customize the generated openapi class name:

```python
import pydantic

class ShortNameClass(pydantic.RootModel):
    root: LongNameClassA | LongNameClassB = Field(
        discriminator="type",
    )


@app.get(
    "/api/my",
    response_model=ShortNameClass,
)
def my_api():
    pass


```
